engine/frontend/src/component/CodeEditor.js

```diff
@@ -0,0 +1,213 @@
+import React, {useEffect, useMemo, useRef, useState} from "react";
+import {Box, Button, useClipboard} from "@chakra-ui/react";
+import useWindowDimensions from "../utils/windowheight";
+import Editor from "@monaco-editor/react";
+
+export const CodeEditor = (
+    dataCallback,
+    readOnly = false,
+    fullFileName = "json_field.json",
+    languages = ["json"],
+    defaultWidthPx = 500,
+    defaultState = languages.includes("json") ? "{\n}" : "",
+    autoFormat = false,
+) => {
+    // https://github.com/microsoft/monaco-editor/blob/main/webpack-plugin/README.md#options
+    const [value, setValue] = useState(defaultState)
+    const contentClipboard = useClipboard("");
+    const monacoRef = useRef(null);
+    const dimensions = useWindowDimensions();
+    const originalReadOnlySetting = useRef(readOnly)
+    const [readOnlySetting, setReadOnlySetting] = useState(readOnly)
+    const [formatCode, setFormatCode] = useState(false)
+
+    // TODO: This could lead to bugs in the future:
+    //  This number depends on the version of Monaco! Use actual enum instead.
+    const monacoReadOnlyEnumId = 86;
+
+    // TODO: Add a promise to getEditor()
+    const getEditor = () => {
+        if (!monacoRef.current) return null;
+        return monacoRef.current.editor.getEditors()[0];
+    }
+
+    const isEditorReadOnly = () => {
+        try {
+            return getEditor().getOption(monacoReadOnlyEnumId)
+        } catch (e) {
+            return undefined
```

Comment:
is it necessary to return undefined ?? also should we instead just log the error message here

Reply:
maybe return false or null.

Reply:
The use case for this to throw an exception is because the editor is not ready yet. In which case I don't want to return false becasue we actually don't know if it's read only or not. We could return null, but I feel undefined captures it better since we don't actually know until the editor is ready. Let me know if you think it's bad practice to use undefined, or I'm overlooking something else!